hadoop-ozone/tools/src/main/java/org/apache/hadoop/ozone/shell/volume/DeleteVolumeHandler.java

```diff
@@ -18,27 +18,220 @@
+        while (iterator.hasNext()) {
+          keys.add(iterator.next().getName());
+          if (MAX_KEY_DELETE_BATCH_SIZE == keys.size()) {
+            bucket.deleteKeys(keys);
+            keys.clear();
+          }
+        }
+        // delete if any remaining keys left
+        if (keys.size() > 0) {
+          bucket.deleteKeys(keys);
+        }
+      }
+      vol.deleteBucket(bucket.getName());
+      numberOfBucketsCleaned.getAndIncrement();
+      return true;
+    } catch (Exception e) {
+      LOG.error("Could not clean bucket ", e);
+      return false;
+    }
+  }
+
+  /**
+   * Clean Legacy/FSO bucket recursively.
+   *
+   * @param  bucket OzoneBucket
+   * @return boolean
+   */
+  private boolean cleanFSBucket(OzoneBucket bucket) {
+    try {
+      String hostPrefix = OZONE_OFS_URI_SCHEME + "://";
+      if (!Strings.isNullOrEmpty(omServiceId)) {
+        hostPrefix += omServiceId + PATH_SEPARATOR_STR;
+      }
+      String ofsPrefix = hostPrefix + vol.getName() + PATH_SEPARATOR_STR +
+          bucket.getName();
+      final Path path = new Path(ofsPrefix);
+      OzoneConfiguration clientConf = new OzoneConfiguration(getConf());
+      clientConf.set(FS_DEFAULT_NAME_KEY, hostPrefix);
+      clientConf.setInt(FS_TRASH_INTERVAL_KEY, 0);
+      FileSystem fs = FileSystem.get(clientConf);
+      if (!fs.delete(path, true)) {
+        throw new IOException("Failed to delete bucket");
+      }
+      numberOfBucketsCleaned.getAndIncrement();
+      return true;
+    } catch (Exception e) {
+      exception = e;
+      LOG.error("Could not clean bucket ", e);
+      return false;
+    }
+  }
+
+  private class BucketCleaner implements Runnable {
+    @Override
+    public void run() {
+      int i;
+      while ((i = cleanedBucketCounter.getAndIncrement()) < totalBucketCount) {
+        try {
+          OzoneBucket bucket = vol.getBucket(bucketIdList.get(i));
+          switch (bucket.getBucketLayout()) {
+          case FILE_SYSTEM_OPTIMIZED:
+          case LEGACY:
+            if (!cleanFSBucket(bucket)) {
+              throw new RuntimeException("Failed to clean bucket");
+            }
+            break;
+          case OBJECT_STORE:
+            if (!cleanOBSBucket(bucket)) {
+              throw new RuntimeException("Failed to clean bucket");
+            }
+          default:
+            throw new RuntimeException("Invalid bucket layout");
+          }
+        } catch (IOException e) {
+          throw new RuntimeException(e);
+        }
+      }
+    }
+  }
+
+  private void doCleanBuckets() throws InterruptedException {
+    int threadPoolSize = 10;
+    executor = Executors.newFixedThreadPool(threadPoolSize);
+    for (int i = 0; i < threadPoolSize; i++) {
+      executor.execute(new BucketCleaner());
+    }
+
+    try {
+      // wait until all Buckets are cleaned or exception occurred.
+      while (numberOfBucketsCleaned.get() != totalBucketCount
```

Comment:
Whats stopping user to create more buckets in between this operation, IIUC, this op doesn't take a global volume lock throughout the recursive volume delete, Say Initially total bucket count = 10  , later I get to create more buckets in b/w as there is some time where the volume/buckets are not locked  , In some time bucket count = 15 , this will exit when deleted buckets = 10 but there are 5 more added. Volume deletes will fail in such a case.

Reply:
Agree with this, this has been already thought and discussed before. When user has decided to delete volume completely, user need to ensure no keys/buckets are getting created in that volume. If that happens then that is user operation problem and volume delete will fail normally. Even if this happen running same command again will delete volume normally.